Mathlib/RingTheory/LocalRing/Module.lean

```diff
@@ -0,0 +1,264 @@
+    rwa [← Submodule.map_comp, this, Submodule.map_top, LinearMap.range_eq_top.mpr hf,
+      LocalRing.map_mkQ_eq_top] at hN
+  · rintro rfl; rw [Submodule.map_top, LinearMap.range_eq_top]
+    exact TensorProduct.mk_surjective R M k Ideal.Quotient.mk_surjective
+
+theorem LocalRing.subsingleton_tensorProduct [Module.Finite R M] :
+    Subsingleton (k ⊗[R] M) ↔ Subsingleton M := by
+  rw [← Submodule.subsingleton_iff R, ← subsingleton_iff_bot_eq_top,
+    ← Submodule.subsingleton_iff R, ← subsingleton_iff_bot_eq_top,
+    ← LocalRing.map_mk_eq_top (M := M), Submodule.map_bot]
+
+theorem LocalRing.span_eq_top_of_tmul_eq_basis [Module.Finite R M] {ι}
+    (f : ι → M) (b : Basis ι k (k ⊗[R] M))
+    (hb : ∀ i, 1 ⊗ₜ f i = b i) : Submodule.span R (Set.range f) = ⊤ := by
+  rw [← LocalRing.map_mk_eq_top, Submodule.map_span, ← Submodule.restrictScalars_span R k
+    Ideal.Quotient.mk_surjective, Submodule.restrictScalars_eq_top_iff,
+    ← b.span_eq, ← Set.range_comp]
+  simp only [Function.comp, mk_apply, hb, Basis.span_eq]
+
+open Function in
+/--
+Given `M₁ → M₂ → M₃ → 0` and `N₁ → N₂ → N₃ → 0`.
+If `M₁ ⊗ N₃ → M₂ ⊗ N₃` and `M₂ ⊗ N₁ → M₂ ⊗ N₂` are both injective,
+then `M₃ ⊗ N₁ → M₃ ⊗ N₂` is also injective.
+-/
+theorem lTensor_injective_of_exact_of_exact_of_rTensor_injective
+    {M₁ M₂ M₃ N₁ N₂ N₃}
+    [AddCommGroup M₁] [Module R M₁] [AddCommGroup M₂] [Module R M₂] [AddCommGroup M₃] [Module R M₃]
+    [AddCommGroup N₁] [Module R N₁] [AddCommGroup N₂] [Module R N₂] [AddCommGroup N₃] [Module R N₃]
+    (f₁ : M₁ →ₗ[R] M₂) (f₂ : M₂ →ₗ[R] M₃) (g₁ : N₁ →ₗ[R] N₂) (g₂ : N₂ →ₗ[R] N₃)
+    (hfexact : Exact f₁ f₂) (hfsurj : Surjective f₂)
+    (hgexact : Exact g₁ g₂) (hgsurj : Surjective g₂)
+    (hfinj : Injective (f₁.rTensor N₃)) (hginj : Injective (g₁.lTensor M₂)) :
+    Injective (g₁.lTensor M₃) := by
+  rw [injective_iff_map_eq_zero]
+  intro x hx
+  obtain ⟨x, rfl⟩ := f₂.rTensor_surjective N₁ hfsurj x
+  have : f₂.rTensor _ (g₁.lTensor _ x) = 0 := by
+    rw [← hx, ← LinearMap.comp_apply, ← LinearMap.comp_apply]
+    congr 1
+    ext x y
+    simp
+  obtain ⟨y, hy⟩ := (rTensor_exact N₂ hfexact hfsurj _).mp this
+  have : g₂.lTensor _ y = 0 := by
+    apply hfinj
+    trans g₂.lTensor _ (g₁.lTensor _ x)
+    · rw [← hy, ← LinearMap.comp_apply, ← LinearMap.comp_apply]
+      congr 1
+      ext x y
+      simp
+    rw [← LinearMap.comp_apply, ← LinearMap.lTensor_comp, hgexact.linearMap_comp_eq_zero]
+    simp
+  obtain ⟨z, rfl⟩ := (lTensor_exact _ hgexact hgsurj _).mp this
+  obtain rfl : f₁.rTensor N₁ z = x := by
+    apply hginj
+    rw [← hy, ← LinearMap.comp_apply, ← LinearMap.comp_apply]
+    congr 1
+    ext x y
+    simp
+  rw [← LinearMap.comp_apply, ← LinearMap.rTensor_comp, hfexact.linearMap_comp_eq_zero]
+  simp
+
+/--
+If `M` is a finitely presented module over a local ring `(R, 𝔪)` such that `m ⊗[R] M → M` is
+injective, then `M` is free.
+-/
+theorem Module.free_of_maximalIdeal_rTensor_injective [Module.FinitePresentation R P]
```

Comment:
Since the proof is a bit long, can you add some comments along it to help readers?